File: scripts/extract-translations.js
const fs = require('fs');
const path = require('path');

/**
 * Extract all text content from story files for translation
 */

const storyDataDir = path.join(__dirname, '../src/data');
const storyFiles = fs
  .readdirSync(storyDataDir)
  .filter(
    file =>
      file.endsWith('.ts') &&
      file.startsWith('story') &&
      !file.includes('_i18n'),
  );

const translationData = {
  metadata: {
    totalStories: 0,
    totalNodes: 0,
    totalChoices: 0,
    generatedAt: new Date().toISOString(),
    instructions:
      'Translate all text fields to target languages. Maintain the structure and IDs. Do not translate IDs, only the text content.',
    targetLanguages: ['ru', 'es', 'de', 'fr', 'pt', 'ja', 'zh', 'ko', 'uk'],
  },
  stories: {},
};

console.log('Extracting translation content from story files...\n');

storyFiles.forEach(file => {
  const filePath = path.join(storyDataDir, file);
  const content = fs.readFileSync(filePath, 'utf8');

  // Extract story ID
  const storyIdMatch = content.match(/id:\s*['"]([^'"]+)['"]/);
  if (!storyIdMatch) {
    console.log(`⚠️  Skipping ${file} - no story ID found`);
    return;
  }

  const storyId = storyIdMatch[1];

  // Extract title
  const titleMatch = content.match(/title:\s*['"]([^'"]+)['"]/);
  const title = titleMatch ? titleMatch[1] : '';

  // Extract description (multiline)
  const descMatch =
    content.match(/description:\s*['"]([^'"]+)['"]/s) ||
    content.match(/description:\s*`([^`]+)`/s);
  const description = descMatch ? descMatch[1].replace(/\\n/g, ' ').trim() : '';

  // Extract author
  const authorMatch = content.match(/author:\s*['"]([^'"]+)['"]/);
  const author = authorMatch ? authorMatch[1] : '';

  // Extract genres
  const genreMatch = content.match(/genre:\s*\[([^\]]+)\]/);
  const genres = genreMatch
    ? genreMatch[1]
        .replace(/['"]/g, '')
        .split(',')
        .map(g => g.trim())
    : [];

  const storyData = {
    storyId,
    sourceFile: file,
    title,
    description,
    author,
    genres,
    nodes: [],
  };

  // Extract all nodes with a more flexible regex
  const nodesSection = content.match(/nodes:\s*\[([\s\S]*)\]\s*,?\s*};?\s*$/m);
  if (nodesSection) {
    const nodesContent = nodesSection[1];

    // Split by node boundaries
    const nodeBlocks = nodesContent.split(/},\s*{/).map((block, index, arr) => {
      if (index === 0) return block + '}';
      if (index === arr.length - 1) return '{' + block;
      return '{' + block + '}';
    });

    nodeBlocks.forEach(block => {
      const nodeIdMatch = block.match(/id:\s*['"]([^'"]+)['"]/);
      const nodeTitleMatch = block.match(/title:\s*['"]([^'"]+)['"]/);
      const narrationMatch =
        block.match(/narration:\s*['"]([^'"]*)['"]/s) ||
        block.match(/narration:\s*`([^`]*)`/s);

      if (!nodeIdMatch || !nodeTitleMatch) return;

      const nodeId = nodeIdMatch[1];
      const nodeTitle = nodeTitleMatch[1];
      const narration = narrationMatch
        ? narrationMatch[1].replace(/\\'/g, "'").replace(/\\n/g, '\n').trim()
        : '';

      const choices = [];

      // Extract choices
      const choicesMatch = block.match(/choices:\s*\[([\s\S]*?)\]/);
      if (choicesMatch) {
        const choicesContent = choicesMatch[1];
        const choiceBlocks = choicesContent.split(/},\s*{/).map((cb, i, a) => {
          if (i === 0 && cb.trim().startsWith('{')) return cb;
          if (i === 0) return '{' + cb + '}';
          if (i === a.length - 1 && cb.trim().endsWith('}')) return cb;
          if (i === a.length - 1) return '{' + cb;
          return '{' + cb + '}';
        });

        choiceBlocks.forEach(cb => {
          const choiceIdMatch = cb.match(/id:\s*['"]([^'"]+)['"]/);
          const choiceTextMatch = cb.match(/text:\s*['"]([^'"]+)['"]/);
          const choiceDescMatch = cb.match(/description:\s*['"]([^'"]+)['"]/);

          if (choiceIdMatch && choiceTextMatch) {
            choices.push({
              choiceId: choiceIdMatch[1],
              text: choiceTextMatch[1].replace(/\\'/g, "'"),
              description: choiceDescMatch
                ? choiceDescMatch[1].replace(/\\'/g, "'")
                : '',
            });
            translationData.metadata.totalChoices++;
          }
        });
      }

      storyData.nodes.push({
        nodeId,
        title: nodeTitle,
        narration,
        choices,
      });

      translationData.metadata.totalNodes++;
    });
  }

  translationData.stories[storyId] = storyData;
  translationData.metadata.totalStories++;

  console.log(`✓ ${file}: ${storyData.nodes.length} nodes`);
});

// Write to file
const outputPath = path.join(__dirname, '../TRANSLATION_SOURCE.json');
fs.writeFileSync(outputPath, JSON.stringify(translationData, null, 2), 'utf8');

console.log(`\n✅ Extraction complete!`);
console.log(`📊 Statistics:`);
console.log(`   - Stories: ${translationData.metadata.totalStories}`);
console.log(`   - Nodes: ${translationData.metadata.totalNodes}`);
console.log(`   - Choices: ${translationData.metadata.totalChoices}`);
console.log(`\n📄 Output: TRANSLATION_SOURCE.json`);
console.log(`\n📝 Next steps:`);
console.log(`   1. Send TRANSLATION_SOURCE.json to translation service`);
console.log(
  `   2. Request translations for languages: ${translationData.metadata.targetLanguages.join(
    ', ',
  )}`,
);
console.log(`   3. Save translated file as TRANSLATION_TARGET.json`);
console.log(`   4. Run: node scripts/import-translations.js`);

File: scripts/extract_translations.py
#!/usr/bin/env python3
"""
Extract all English text from story files for translation purposes.
Creates a comprehensive JSON file with all titles, descriptions, narrations, and choices.
"""

import json
import re
import os
from pathlib import Path

def extract_string_value(content, pattern):
    """Extract string value from TypeScript code."""
    match = re.search(pattern, content, re.DOTALL)
    if match:
        value = match.group(1)
        # Clean up the string
        value = value.replace('\\n', '\n')
        value = value.replace("\\'", "'")
        value = value.replace('\\"', '"')
        return value.strip()
    return None

def extract_story_metadata(content):
    """Extract story metadata (title, author, description)."""
    metadata = {}
    
    # Extract title
    title = extract_string_value(content, r"title:\s*['\"]([^'\"]+)['\"]")
    if title:
        metadata['title'] = title
    
    # Extract author
    author = extract_string_value(content, r"author:\s*['\"]([^'\"]+)['\"]")
    if author:
        metadata['author'] = author
    
    # Extract description
    desc = extract_string_value(content, r"description:\s*['\"]([^'\"]+)['\"]")
    if desc:
        metadata['description'] = desc
    
    return metadata

def extract_nodes(content):
    """Extract all nodes with their narrations and choices."""
    nodes = {}
    
    # Find all node definitions
    node_pattern = r"\{\s*id:\s*['\"]([^'\"]+)['\"],[^}]+title:\s*['\"]([^'\"]+)['\"],[^}]+narration:\s*['\"]([^'\"]+)['\"]"
    
    for match in re.finditer(node_pattern, content, re.DOTALL):
        node_id = match.group(1)
        title = match.group(2)
        narration = match.group(3)
        
        # Clean up narration
        narration = narration.replace('\\n', '\n').replace("\\'", "'").replace('\\"', '"')
        
        nodes[node_id] = {
            'title': title,
            'narration': narration,
            'choices': {}
        }
    
    # Extract choices for each node
    choice_pattern = r"\{\s*id:\s*['\"]([^'\"]+)['\"],[^}]+text:\s*['\"]([^'\"]+)['\"](?:,[^}]+description:\s*['\"]([^'\"]*)['\"])?"
    
    for match in re.finditer(choice_pattern, content, re.DOTALL):
        choice_id = match.group(1)
        text = match.group(2)
        description = match.group(3) if match.group(3) else None
        
        # Find which node this choice belongs to (approximate)
        # This is a simplified approach
        for node_id in nodes:
            if choice_id.startswith(node_id.split('-')[0]):
                nodes[node_id]['choices'][choice_id] = {
                    'text': text,
                    'description': description
                }
                break
    
    return nodes

def process_story_file(filepath):
    """Process a single story file and extract all translatable content."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    story_data = {}
    
    # Extract metadata
    metadata = extract_story_metadata(content)
    story_data.update(metadata)
    
    # Extract nodes
    nodes = extract_nodes(content)
    if nodes:
        story_data['nodes'] = nodes
    
    return story_data

def main():
    """Main extraction function."""
    data_dir = Path('src/data')
    output_file = 'STORY_TRANSLATIONS_EN_COMPLETE.json'
    
    all_stories = {}
    
    # Process all story files
    story_files = list(data_dir.glob('story*.ts')) + list(data_dir.glob('*Stories.ts'))
    
    for story_file in sorted(story_files):
        print(f"Processing {story_file.name}...")
        try:
            story_data = process_story_file(story_file)
            if story_data:
                # Use filename as key
                story_key = story_file.stem
                all_stories[story_key] = story_data
        except Exception as e:
            print(f"Error processing {story_file.name}: {e}")
    
    # Create final JSON structure
    output = {
        "meta": {
            "version": "1.0",
            "language": "en",
            "description": "Complete English translations for all story content",
            "lastUpdated": "2025-01-15",
            "totalStories": len(all_stories)
        },
        "stories": all_stories
    }
    
    # Write to file
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(output, f, indent=2, ensure_ascii=False)
    
    print(f"\nExtraction complete! Created {output_file}")
    print(f"Total stories processed: {len(all_stories)}")

if __name__ == '__main__':
    main()
